File: day2.ml
[@@@warning "-8"]
let input = List.map (fun [a; b] -> (a, b)) (
    List.map (String.split_on_char '-') (
        String.split_on_char ',' (String.trim (In_channel.with_open_text "./day2.txt" In_channel.input_all))
        (* String.split_on_char ',' (String.trim (In_channel.with_open_text "./test.txt" In_channel.input_all)) *)
        )
    )
[@@@warning "+8"]

let part1 input = 
    let check_num num_str =
        if String.length num_str mod 2 = 0 then
            if (String.sub num_str 0 (String.length num_str / 2)) = (String.sub num_str (String.length num_str / 2) (String.length num_str / 2))
            then int_of_string num_str
            else 0
        else 0
    in let rec range num target =
        if int_of_string num > int_of_string target then 0 else check_num num + range (string_of_int (int_of_string num +1)) target
    in let rec loop lst =
        match lst with
        | head :: tail -> range (fst head) (snd head) + loop tail
        | [] -> 0
    in loop input

let part2 input = 
    let rec all_equal prev lst =
        match lst with
        | first :: tail -> (first = prev) && (all_equal first tail)
        | [] -> true
    in let rec split_str num str =
        if String.length str = 0 then []
        else String.sub str 0 num :: split_str num (String.sub str num (String.length str - num))
    in let rec repeats num num_str =
        if num = String.length num_str then 0
        else 
            if String.length num_str mod num = 0 then 
                if all_equal (String.sub num_str 0 num) (split_str num num_str) then
                    int_of_string num_str
                else repeats (num+1) num_str
            else repeats (num+1) num_str
    in let rec range num target =
        if int_of_string num > int_of_string target then 0 else repeats 1 num + range (string_of_int (int_of_string num +1)) target
    in let rec loop lst =
        match lst with
        | head :: tail -> range (fst head) (snd head) + loop tail
        | [] -> 0
    in loop input

let time f x =
    let t = Sys.time() in
    let fx = f x in
    Printf.printf "%d (%fs)\n%!" fx (Sys.time() -. t)

let () = time part1 input
let () = time part2 input

File: day2.py
import time

with open("day2.txt") as txt:
    input = txt.read().strip().split(",")
    input = [i.split("-") for i in input]

def part1():
    count = 0
    for i in input:
        for j in range(int(i[0]), int(i[1])+1):
            j = str(j)
            if j[:len(j)//2] == j[len(j)//2:]:
                count += int(j)
    return count

def split(num, splits):
    if len(num) % splits == 0:
        seg_len = len(num) // splits
        arr = []
        for i in range(0, len(num), seg_len):
            arr.append(num[i:i+seg_len])
        return arr

def part2():
    count = 0
    for i in input:
        for j in range(int(i[0]), int(i[1])+1):
            j = str(j)
            for k in range(2, len(i[1])+1):
                splits = split(j, k)
                if splits is not None and len(set(splits)) == 1:
                    count += int(j)
                    break
    return count

start = time.time()
print(part1(), "(%fs)" % (time.time()-start))
start = time.time()
print(part2(), "(%fs)" % (time.time()-start))

